server/src/main/java/com/microsoft/java/bs/core/internal/reporter/ProgressReporter.java

```diff
@@ -3,32 +3,71 @@
 
 package com.microsoft.java.bs.core.internal.reporter;
 
-import ch.epfl.scala.bsp4j.StatusCode;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+import org.gradle.tooling.events.OperationDescriptor;
+import org.gradle.tooling.events.ProgressListener;
+import org.gradle.tooling.events.task.TaskOperationDescriptor;
+
+import ch.epfl.scala.bsp4j.BuildClient;
+import ch.epfl.scala.bsp4j.LogMessageParams;
+import ch.epfl.scala.bsp4j.MessageType;
+import ch.epfl.scala.bsp4j.TaskId;
 
 /**
- * A progress reporter that reports the progress of a task.
+ * An extension of {@link ProgressListener} that allows sending errors.
  */
-public interface ProgressReporter {
+public abstract class ProgressReporter implements ProgressListener {
 
-  /**
-   * Notify the client that a task has been started.
-   *
-   * @param message the message to be displayed.
-   */
-  void taskStarted(String message);
+  protected final BuildClient client;
+  protected final TaskId taskId;
+  protected final List<String> taskIds;
 
   /**
-   * Notify the progress of the task.
+   * Instantiates a {@link ProgressReporter}.
    *
-   * @param message the message to be displayed.
+   * @param client BSP client to report to.
    */
-  void taskInProgress(String message);
+  public ProgressReporter(BuildClient client) {
+    this.client = client;
+    taskId = new TaskId(UUID.randomUUID().toString());
+    taskIds = new ArrayList<>();
+    taskIds.add(taskId.getId());
+  }
 
   /**
-   * Notify the client that a task has been finished.
+   * Notify the client of an error.
    *
-   * @param message the message to be displayed.
-   * @param statusCode the status code of the task.
+   * @param error the error message.
    */
-  void taskFinished(String message, StatusCode statusCode);
+  public void sendError(String error) {
+    if (client != null) {
+      LogMessageParams messageParam = new LogMessageParams(MessageType.ERROR, error);
+      messageParam.setTask(taskId);
+      client.onBuildLogMessage(messageParam);
+    }
+  }
+
+  protected TaskId getTaskId(String taskPath) {
+    TaskId taskId = new TaskId(taskPath == null ? "null" : taskPath);
+    taskId.setParents(taskIds);
```

Comment:
May I ask what's the purpose of setting parents here? I'm a little bit confused here.

Reply:
It's just for debugging reference.  To be able to track down which messages belong to which tasks.

E.g. The BSP client might trigger a compilation for the `model` build target.
The build server would run the `classes` task.
The `ProgressReporter` would create a unique taskId e.g. `2b4287e5-71a8-47e9-9202-124dcf0a052c`
Then messages would be sent with whatever Gradle task name is and with the unique Id as the parent...

```json
[Trace - 07:41:04 pm] Received notification 'build/taskStart'
Params: {
  "taskId": {
    "id": ":model:compileJava",
    "parents": [
      "2b4287e5-71a8-47e9-9202-124dcf0a052c"
    ]
  }
}
[Trace - 07:41:05 pm] Received notification 'build/taskFinish'
Params: {
  "taskId": {
    "id": ":model:compileJava",
    "parents": [
      "2b4287e5-71a8-47e9-9202-124dcf0a052c"
    ]
  }
}
[Trace - 07:41:05 pm] Received notification 'build/taskStart'
Params: {
  "taskId": {
    "id": ":model:processResources",
    "parents": [
      "2b4287e5-71a8-47e9-9202-124dcf0a052c"
    ]
  }
}
[Trace - 07:41:05 pm] Received notification 'build/taskFinish'
Params: {
  "taskId": {
    "id": ":model:processResources",
    "parents": [
      "2b4287e5-71a8-47e9-9202-124dcf0a052c"
    ]
  }
}
[Trace - 07:41:05 pm] Received notification 'build/taskStart'
Params: {
  "taskId": {
    "id": ":model:classes",
    "parents": [
      "2b4287e5-71a8-47e9-9202-124dcf0a052c"
    ]
  }
}
[Trace - 07:41:05 pm] Received notification 'build/taskFinish'
Params: {
  "taskId": {
    "id": ":model:classes",
    "parents": [
      "2b4287e5-71a8-47e9-9202-124dcf0a052c"
    ]
  }
}
```
I've cut off the body of the message to just highlight the `taskId` section.
You can see that Gradle is executing `compileJava`, `processResources`, `classes` for the `model` target and that all those progress reports relate to the original `2b4287e5-71a8-47e9-9202-124dcf0a052c` task.
Without using `parents` you would just see `2b4287e5-71a8-47e9-9202-124dcf0a052c` task.  You could examine the `message` to see what's going on but I think this makes for easier debugging.

Reply:
I see. Thanks for the explanation. I'm clear now.